src/tools/rbd/action/Ubbd.cc

```diff
@@ -0,0 +1,190 @@
+
+int execute_list(const po::variables_map &vm,
+                 const std::vector<std::string> &ceph_global_init_args) {
+#if defined(WITH_RBD_UBBD)
+  struct ubbdd_mgmt_rsp rsp = { 0 };
+  struct ubbd_list_options list_opts = { .type = UBBD_DEV_TYPE_RBD };
+  int r;
+  int i;
+
+  r = ubbd_list(&list_opts, &rsp);
+  if (r)
+    return r;
+
+  for (i = 0; i < rsp.list.dev_num; i++) {
+    std::cout << "/dev/ubbd" << rsp.list.dev_list[i] << std::endl;
+  }
+
+  return 0;
+#else
+  std::cerr << "rbd: ubbd device is not supported" << std::endl;
+  return -EOPNOTSUPP;
+#endif
+}
+
+int execute_map(const po::variables_map &vm,
+                const std::vector<std::string> &ceph_global_init_args) {
+#if defined(WITH_RBD_UBBD)
+  size_t arg_index = 0;
+  std::string pool_name;
+  std::string nspace_name;
+  std::string image_name;
+  std::string snap_name;
+  struct ubbdd_mgmt_rsp rsp = { 0 };
+  struct ubbd_map_options opts = { 0 };
+  std::string dev_type = "rbd";
+
+  int r = utils::get_pool_image_snapshot_names(
+    vm, at::ARGUMENT_MODIFIER_NONE, &arg_index, &pool_name, &nspace_name,
+    &image_name, &snap_name, true, utils::SNAPSHOT_PRESENCE_PERMITTED,
+    utils::SPEC_VALIDATION_NONE);
+  if (r < 0) {
+    return r;
+  }
+  opts.type = opts.generic_dev.opts.type = dev_type.c_str();
+  opts.generic_dev.opts.rbd.pool = pool_name.c_str();
+  opts.generic_dev.opts.rbd.ns = nspace_name.c_str();
+  opts.generic_dev.opts.rbd.image = image_name.c_str();
+  opts.generic_dev.opts.rbd.snap = snap_name.c_str();
+
+  if (vm["read-only"].as<bool>()) {
+    opts.read_only = true;
+  }
+  if (vm["exclusive"].as<bool>()) {
+	  opts.generic_dev.opts.rbd.exclusive = true;
+  }
+  if (vm["quiesce"].as<bool>()) {
+	  opts.generic_dev.opts.rbd.quiesce = true;
+  }
+  if (vm.count("quiesce-hook")) {
+	  opts.generic_dev.opts.rbd.quiesce_hook = vm["quiesce-hook"].as<std::string>().c_str();
+  }
+
+  r = ubbd_map(&opts, &rsp);
+  if (r < 0) {
+    std::cerr << "rbd: ubbd device map failed: " << cpp_strerror(r) << std::endl;
+    return r;
+  }
+
+  std::cout << rsp.add.path << std::endl;
+
+  return 0;
+#else
+  std::cerr << "rbd: ubbd device is not supported" << std::endl;
+  return -EOPNOTSUPP;
+#endif
+}
+
+#if defined(WITH_RBD_UBBD)
+static int parse_unmap_options(const std::string &options_string,
+		struct ubbd_unmap_options *unmap_opts)
+{
+  char *options = strdup(options_string.c_str());
+  BOOST_SCOPE_EXIT(options) {
+    free(options);
+  } BOOST_SCOPE_EXIT_END;
+
+  for (char *this_char = strtok(options, ", ");
+       this_char != NULL;
+       this_char = strtok(NULL, ",")) {
+    char *value_char;
+
+    if ((value_char = strchr(this_char, '=')) != NULL)
+      *value_char++ = '\0';
+
+    if (!strcmp(this_char, "force")) {
+      unmap_opts->force = true;
+    } else {
+      std::cerr << "rbd: ubbd unknown unmap option '" << this_char << "'"
+                << std::endl;
+      return -EINVAL;
+    }
+  }
+
+  return 0;
+}
```

Comment:
nit: I would have `#endif` here just to make `execute_unmap` consistent with other functions.

Reply:
Thanx, but the parse_unmap_options only exist in `#if defined(WITH_RBD_UBBD)`, then we can see:
```
#if defined(WITH_RBD_UBBD)
    parse_unmap_options()
    execute_unmap()
#else 
    execute_unmap()
#endif
```

do you mean:
```
#if defined(WITH_RBD_UBBD)
    parse_unmap_options()
#endif

#if defined(WITH_RBD_UBBD)
    execute_unmap()
#else 
    execute_unmap()
#endif
```

like this?

Reply:
I mean:
```
#if defined(WITH_RBD_UBBD)
parse_unmap_options()
#endif

execute_unmap() {
#if defined(WITH_RBD_UBBD)
   ...
#else 
  std::cerr << "rbd: ubbd device is not supported" << std::endl;
  return -EOPNOTSUPP;
#endif
}
```